bee-vote/src/fpc.rs

```diff
@@ -0,0 +1,587 @@
+                        object: *object,
+                        opinion: context.last_opinion().ok_or(Error::Unknown("No opinions found"))?,
+                        context: context.clone(),
+                    }))
+                    .or(Err(Error::SendError))?;
+
+                to_remove.push(*object);
+            }
+        }
+
+        // Remove any finalized/failed votes.
+        for object in to_remove {
+            context_guard.remove(&object);
+        }
+
+        Ok(())
+    }
+
+    /// Perform the voting round, with a given threshold (between 0 and 1).
+    /// This threshold is used to generate opinions on the voting object.
+    ///
+    /// For each `VoteContext` in the voting pool, a random number is generated within the range
+    /// given on initialisation of the `Fpc` struct, and compared to the threshold to generate a
+    /// `Like` or `Dislike` opinion.
+    pub async fn do_round(&self, rand: f64) -> Result<(), Error> {
+        let start = Instant::now();
+        self.enqueue().await;
+
+        if self.last_round_successful.load(Ordering::Relaxed) {
+            self.form_opinions(rand).await;
+            self.finalize_opinions().await?;
+        }
+
+        let queried_opinions = self.query_opinions().await?;
+        self.last_round_successful.store(true, Ordering::Relaxed);
+
+        let round_stats = RoundStats {
+            duration: start.elapsed(),
+            rand_used: rand,
+            vote_contexts: self.contexts.read().await.clone(),
+            queried_opinions,
+        };
+
+        self.tx
+            .send(Event::RoundExecuted(round_stats))
+            .or(Err(Error::SendError))?;
+
+        Ok(())
+    }
+
+    /// Select a number of `OpinionGiver`s and query them for opinions.
+    async fn query_opinions(&self) -> Result<Vec<QueriedOpinions>, Error> {
+        let mut rng = thread_rng();
+
+        let query_ids = self.vote_context_ids().await;
+        if query_ids.conflict_objects.is_empty() && query_ids.timestamp_objects.is_empty() {
+            return Ok(vec![]);
+        }
+
+        // Create opinion givers.
+        let mut opinion_givers = (self.opinion_giver_fn)();
+        if opinion_givers.is_empty() {
+            return Err(Error::NoOpinionGivers);
+        }
+
+        let dist = rand::distributions::Uniform::new(0, opinion_givers.len());
+
+        // Select queries.
+        let mut queries = vec![0u32; opinion_givers.len()];
+        for _ in 0..self.query_sample_size {
+            let index = rng.sample(dist);
+
+            if let Some(selected_count) = queries.get_mut(index) {
+                *selected_count += 1;
+            }
+        }
+
+        let vote_map = Arc::new(RwLock::new(HashMap::<VoteObject, Opinions>::new()));
+        let all_queried_opinions = Arc::new(RwLock::new(Vec::<QueriedOpinions>::new()));
+        let mut query_futures = vec![];
+
+        for (i, opinion_giver) in opinion_givers.iter_mut().enumerate() {
```

Comment:
Second approach sounds best here 👍

Reply:
Settled with the first approach for now actually, because I feel it's a little easier to read and reason around (especially since this code is pretty opaque anyway, and I don't really want to introduce another struct).